### DISCUSSION TITLE
Roadmapping for Tesseract

### DISCUSSION THREAD

**Discussion by rjoomen** [tesseract-robotics/tesseract#948] [Ideas]
I'm starting a new discussion to talk about @marip8's [roadmapping idea](https://github.com/tesseract-robotics/tesseract/discussions/944), so as to not clutter the roadmap itself. 

I think a more structured development approach is a good idea, like @marip8 already mentioned in the last ROS-I developer's meeting, but I have some remarks:

1) Is this going to be a repo-specific roadmap? You current proposal is quite detailed and involves the `tesseract` repo only. I think what's also needed is an overarching roadmap for the entire `tesseract-robotics` organization, where larger goals (like the profile refactor, the completion of trajopt_ifopt, Tesseract Studio etc.) are listed.
2) I also think a Tesseract developers communications channel/meetings or something like that would be a good initiative to sync the work done by separate people and institutions.

@Levi-Armstrong, @marrts, @johnwason, @mpowelson what do you think?

**Comment by marip8** [comment]
> I think what's also needed is an overarching roadmap for the entire tesseract-robotics organization, where larger goals (like the profile refactor, the completion of trajopt_ifopt, Tesseract Studio etc.) are listed.

I think this is a good idea as well. I think this would be the high-level description of what the upcoming changes are and why they are needed; then we could potentially use per-repo Github Projects to define the specific tasking. I'm not entirely sure where this tesseract-org level roadmap would go though. Maybe we move it to the `tesseract_docs` repo? Or we could start a new `tesseract-roadmap` repo just for discussions and other high level documentation/diagrams, etc.

**Comment by johnwason** [comment]
I will be developing tesseract_robotraconteur regardless. I started working on it previously, but did not finish it. I am busy working on a few other projects, but once they are done I will be focusing on tesseract_robotraconteur.

These are incomplete, but I created a gist of some concept Robot Raconteur service definitions for the updated tesseract_robotraconteur: https://gist.github.com/johnwason/9570213e62fb737162170c53194cbb05

Robot Raconteur allows for a much closer interface to the C++ API, and has other capabilities like plug-and-play and security that make it capable and easy to use.

**Reply by johnwason** [reply]
Here is an example of the interface for Gazebo: https://github.com/robotraconteur-contrib/RobotRaconteur_Gazebo_Server_Plugin/blob/github/gazebo_robotraconteur_server_plugin/org.gazebosim.gazebo.robdef

Here are the general standardized interfaces: https://github.com/robotraconteur/robotraconteur_standard_robdef

**Reply by rjoomen** [reply]
I'm confused, is this in reply to the present roadmapping discussion, or to https://github.com/tesseract-robotics/tesseract_ros2/issues/55?

**Reply by johnwason** [reply]
Are we not discussing high level roadmapping? This repo will be moved to tesseract-robotics once it is working.

**Reply by marip8** [reply]
I think we should define the bounds of our roadmap(s) first. IMO it makes sense to group the "core" `tesseract` repos (`tesseract`, `tesseract_planning`, `tesseract_qt` (excluding studio?), `tesseract_python`) into one roadmap and then maintain individual roadmaps for projects that bring `tesseract` into other frameworks (e.g., ROS, ROS2, robot raconteur, gazebo, tesseract studio?) in their own repos. I think having a single roadmap for the entire `tesseract` ecosystem is probably too large and messy.

**Reply by johnwason** [reply]
I also started working on a C# interface to tesseract: https://github.com/johnwason/tesseract_csharp . I don't have funding for this work so I haven't made much progress. It may be worth considering since a lot of industrial systems use C#.

**Reply by rjoomen** [reply]
@johnwason you're right, this fits the larger roadmap (I was thrown off by 'regardless').

**Reply by johnwason** [reply]
@rjoomen to be clear, tesseract_robotraconteur and tesseract_csharp are different subprojects. I mentioned tesseract_csharp because @marip8 mentioned tesseract_python, and tesseract_csharp is an offshoot of tesseract_python taking advantage of the work with SWIG. A lot of industrial software uses C# so it would be valuable to include in the roadmap. tesseract_robotraconteur is a critical component of the Robot Raconteur ecosystem so I will be working on it as soon as I complete some other work.

**Comment by rjoomen** [comment]
Adding the roadmap to `tesseract_docs` is a good idea. I took a quick look at two large ROS projects, [MoveIt](https://moveit.ros.org/documentation/contributing/roadmap/) and [Nav2](https://navigation.ros.org/roadmap/roadmap.html) and they have fairly extensive roadmaps published with their docs. (Their docs are also better, a good point for our roadmap.)

**Comment by johnwason** [comment]
GitHub has a pretty good milestone system where you can assign issues to a specific version release/milestone. It is used for example in the csharplang repository to work on the development of the programming language and seems to work well.

https://github.com/dotnet/csharplang/milestones
